### PR TITLE
Set Counterparty Last Sequence Send

### PR DESCRIPTION
**Pull request by chatton**
## Description

Note: in the spec this happens in the startFlushHandshake function, but since our channel writes happen separately, it needs to be added in all places that write the channel. This is a potential gotcha.

closes: #4001

### Commit Message / Changelog Entry

N/A Feature branch

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Review by charleenfei — APPROVED** [review]
nice work! lgtm

File: modules/core/04-channel/keeper/upgrade.go [hunk 1]
```diff
@@ -199,6 +199,7 @@ func (k Keeper) WriteUpgradeTryChannel(ctx sdk.Context, portID, channelID string
 
 	upgrade.Fields.Version = upgradeVersion
 
+	k.SetCounterpartyLastPacketSequence(ctx, portID, channelID, upgrade.LatestSequenceSend)
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade.go:202]: this is our `Upgrade`'s last sequence send, right? I think we'll need to accept an additional arg and pass `msg.CounterpartyProposedUpgrade.LastSequenceSend` from msg server

Comment (chatton) [reply modules/core/04-channel/keeper/upgrade.go:202]: yes! Great catch, we'll need to add that as an arg.

File: modules/core/04-channel/keeper/upgrade.go [hunk 2]
```diff
@@ -181,7 +181,7 @@ func (k Keeper) ChanUpgradeTry(
 
 // WriteUpgradeTryChannel writes the channel end and upgrade to state after successfully passing the UpgradeTry handshake step.
 // An event is emitted for the handshake step.
-func (k Keeper) WriteUpgradeTryChannel(ctx sdk.Context, portID, channelID string, upgrade types.Upgrade, upgradeVersion string) (types.Channel, types.Upgrade) {
+func (k Keeper) WriteUpgradeTryChannel(ctx sdk.Context, portID, channelID string, upgrade types.Upgrade, upgradeVersion string, counterpartyLastPacketSequence uint64) (types.Channel, types.Upgrade) {
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/upgrade.go:184]: small nit: should we make the args consistent across both write funcs. maybe `counterpartyLastSequenceSend`

**Review by damiannolan — APPROVED** [review]

File: modules/core/04-channel/keeper/upgrade.go (hunk 2 above)
Comment (chatton) [reply modules/core/04-channel/keeper/upgrade.go]: we most certainly can!

**Pull request merged**